src/shared/pam-util.c

```diff
@@ -21,6 +23,32 @@ int pam_syslog_errno(pam_handle_t *handle, int level, int error, const char *for
         return error == -ENOMEM ? PAM_BUF_ERR : PAM_SERVICE_ERR;
 }
 
+int pam_syslog_pam_error(pam_handle_t *handle, int level, int error, const char *format, ...) {
+        /* This wraps pam_syslog() but will replace @PAMERR@ with a string from pam_strerror().
+         * @PAMERR@ must be at the very end. */
+
+        va_list ap;
+        va_start(ap, format);
+
+        const char *p = endswith(format, "@PAMERR@");
+        if (p) {
+                const char *pamerr = pam_strerror(handle, error);
+                if (strchr(pamerr, '%'))
+                        pamerr = "n/a";  /* We cannot have any formatting chars */
```

Comment:
maybe just use `strreplace("%", "%%")`? or

Reply:
We *could* do this, but it'd require some special logic to handle longer strings. I think the issue is purely theoretical — I don't see why an error message would ever have a percent character. I don't think it's worth making the code more complex to cater to such a hypothetical scenario.